examples/optimization.py

```diff
@@ -14,23 +11,27 @@
 tf_phi = tf.Variable(0.1)
 
 alpha, phi = circuit.params('alpha', 'phi')
+
 with circuit.context as q:
     Dgate(alpha, phi) | q[0]
-results = eng.run(circuit, args={'alpha': tf_alpha, 'phi': tf_phi}, eval=False})
 
-# loss is probability for the coherent state |alpha=1>
-prob = results.state.fidelity_coherent([1.0])
-loss = -prob  # negative sign to maximize prob
+opt = tf.keras.optimizers.Adam(learning_rate=0.1)
+steps = 50
+
+for step in range(steps):
 
-# Set up optimization
-optimizer = tf.train.GradientDescentOptimizer(learning_rate=0.1)
-minimize_op = optimizer.minimize(loss)
+    # reset the engine if it has already been executed
+    if eng.run_progs:
+        eng.reset()
 
-# Create Tensorflow Session and initialize variables
-sess = tf.Session()
-sess.run(tf.global_variables_initializer())
+    with tf.GradientTape() as tape:
+        # execute the engine
+        results = eng.run(circuit, args={'alpha': tf_alpha, 'phi': tf_phi})
+        # get the probability of fock state |1>
+        prob = results.state.fock_prob([1])
+        # negative sign to maximize prob
+        loss = -prob
 
-# Carry out optimization
-for step in range(50):
-    prob_val, _ = sess.run([prob, minimize_op])
-    print("Value at step {}: {}".format(step, prob_val))
+    gradients = tape.gradient(loss, [tf_alpha, tf_phi])
+    opt.apply_gradients(zip(gradients, [tf_alpha, tf_phi]))
```

Comment:
Could [`tf.minimize`](https://github.com/tensorflow/tensorflow/blob/e5bf8de410005de06a7ff5393fafdf832ef1d4ad/tensorflow/python/keras/optimizer_v2/optimizer_v2.py#L289) be used here for brevity?

Based on the docstrings, it seems that the `tf.GradientTape()` + `apply_gradients` method approach is useful when the `gradients` are being processed (I might be missing something here!).

Reply:
Oh good catch! This didn't exist back when I first started using eager mode 😆

Reply:
Have added a section using `tf.minimize` :) Note that we can't fully replace the existing approach, as `tf.minimize` doesn't allow us to extract intermediate values (such as the probability), but it's nice to show both methods.